### ISSUE TITLE
Custom Product Attribute changes 'backend_type' when 'is_user_defined = 1' and get updated/saved in Admin Backend

### ISSUE THREAD

**Issue by mhauri** [magento/magento2#9219]
### Preconditions
1. Magento: 2.1.4 to 2.1.6, maybe earlier versions are affected too.

### Steps to reproduce
1. Create a custom Product attribute via Setup of a Custom Module:
```
[
    'required' => 0,
    'global' => ScopedAttributeInterface::SCOPE_GLOBAL,
    'group' => 'Test',
    'searchable' => 0,
    'comparable' => 0,
    'used_for_sort_by' => 0,
    'used_in_product_listing' => 0,
    'visible_in_advanced_search' => 0,
    'user_defined' => 1,
    'is_html_allowed_on_front' => 1,
    'visible_on_front' => 1,
    'input' => 'select',
    'source' => 'Test\Test\Model\Attribute\Source\Discount',
    'backend_type' => 'varchar',
    'label' => 'Discount',
    'position' => 1,
]
```
2. Go to the Backend **Stores** > **Attributes** > **Product**
3. Click on the newly created attribute and save it (with or without modifications).

### Expected result

1. The backend_type doesn't change for the custom attribute.

### Actual result
1. The backend_type got changed from **varchar** to **int**

**Comment by mhauri** [comment]
After digging into the code the problem starts in the file: _module-catalog/Controller/Adminhtml/Product/Attribute/Save.php_ on line 199

``` 
if (is_null($model->getIsUserDefined()) || $model->getIsUserDefined() != 0) {
    $data['backend_type'] = $model->getBackendTypeByInput($data['frontend_input']);
}
```

So when the Attribute is user defined `$model->getIsUserDefined() != 0` then the backend_type will be set based on the frontend_input `$model->getBackendTypeByInput($data['frontend_input'])`

The method _getBackendTypeByInput()_ is defined in: _module-eav/Model/Entity/Attribute.php_

```
/**
     * Detect backend storage type using frontend input type
     *
     * @param string $type frontend_input field value
     * @return string backend_type field value
     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
     */
    public function getBackendTypeByInput($type)
    {
        $field = null;
        switch ($type) {
            case 'text':
            case 'gallery':
            case 'media_image':
            case 'multiselect':
                $field = 'varchar';
                break;

            case 'image':
            case 'textarea':
                $field = 'text';
                break;

            case 'date':
                $field = 'datetime';
                break;

            case 'select':
            case 'boolean':
                $field = 'int';
                break;

            case 'price':
            case 'weight':
                $field = 'decimal';
                break;

            default:
                break;
        }

        return $field;
    }
```

By this method the frontend_input: _select_ will always return _int_
As far as I can see there are core attributes which also have _select_ as frontend_input and _varchar_ as backend_type.

For the moment the solution will be to change the custom attribute to _is_user_defined = 0_ but I think this should be handled different for user defined attributes by not setting the _backend_type_ based on the _frontend_input_ 

Or are there any reasons this is handled this way?

**Comment by magento-engcom-team** [comment]
@mhauri, thank you for your report.
We've created internal ticket(s) MAGETWO-80862 to track progress on the issue.

**Comment by keyurshah070** [comment]
should I commit ? After checking long conversion I am confuse(whether it is merged or not) in status @okorshenko @ishakhsuvarov

**Comment by keyurshah070** [comment]
I'm working on it #SQUASHTOBERFEST @okorshenko @ishakhsuvarov

**Comment by sidolov** [comment]
Hi @mhauri. Thank you for your report. 
The issue has been fixed in magento/magento2#18196 by @bartoszkubicki in 2.2-develop branch
Related commit(s):
 - [f6a94595ace124761b9c423f7655b63db48e6545](https://github.com/magento/magento2/commit/f6a94595ace124761b9c423f7655b63db48e6545)

The fix will be available with the upcoming 2.2.8 release.

**Comment by slavvka** [comment]
Hi @mhauri. Thank you for your report. 
The issue has been fixed in magento/magento2#18570 by @gelanivishal in 2.3-develop branch
Related commit(s):
 - [40e128c14e32e5005d35a79e0dc7936728414b0e](https://github.com/magento/magento2/commit/40e128c14e32e5005d35a79e0dc7936728414b0e)

The fix will be available with the upcoming 2.3.1 release.
